src/main/java/com/facebook/presto/pinot/PinotScatterGatherQueryClient.java

```diff
@@ -0,0 +1,397 @@
+            ScatterGatherStats scatterGatherStats, boolean isOfflineTable,
+            String tableNameWithType)
+    {
+        try {
+            Map<ServerInstance, byte[]> serverResponseMap = compositeFuture.get();
+            if (!ignoreEmptyResponses) {
+                if (serverResponseMap.size() != routingTable.size()) {
+                    Map<String, String> routingTableForLogging = new HashMap<>();
+                    routingTable.entrySet().forEach(entry -> {
+                        String valueToPrint = entry.getValue().size() > 10 ? format("%d segments", entry.getValue().size()) : entry.getValue().toString();
+                        routingTableForLogging.put(entry.getKey(), valueToPrint);
+                    });
+                    throw new PinotException(ErrorCode.PINOT_INSUFFICIENT_SERVER_RESPONSE, String.format("%d of %d servers responded with routing table servers: %s, error: %s", serverResponseMap.size(), routingTable.size(), routingTableForLogging, compositeFuture.getError()));
+                }
+                Iterator<Map.Entry<ServerInstance, byte[]>> iterator = serverResponseMap.entrySet().iterator();
+                while (iterator.hasNext()) {
+                    Map.Entry<ServerInstance, byte[]> entry = iterator.next();
+                    if (entry.getValue().length == 0) {
+                        throw new PinotException(ErrorCode.PINOT_INSUFFICIENT_SERVER_RESPONSE, String.format("Got empty response with from server: %s", entry.getKey().getShortHostName()));
+                    }
+                }
+            }
+            Map<ServerInstance, Long> responseTimes = compositeFuture.getResponseTimes();
+            scatterGatherStats.setResponseTimeMillis(responseTimes, isOfflineTable);
+            return serverResponseMap;
+        }
+        catch (ExecutionException | InterruptedException e) {
+            Throwable err = e instanceof ExecutionException ? ((ExecutionException) e).getCause() : e;
+            throw new PinotException(ErrorCode.PINOT_UNCLASSIFIED_ERROR, String.format("Caught exception while fetching responses for table: %s", tableNameWithType), err);
+        }
+    }
+
+    /**
+     * Deserialize the server responses, put the de-serialized data table into the data table map passed in, append
+     * processing exceptions to the processing exception list passed in.
+     * <p>For hybrid use case, multiple responses might be from the same instance. Use response sequence to distinguish
+     * them.
+     *
+     * @param responseMap map from server to response.
+     * @param isOfflineTable whether the responses are from an OFFLINE table.
+     * @param tableNameWithType table name with type suffix.
+     * @return dataTableMap map from server to data table.
+     */
+    private Map<ServerInstance, DataTable> deserializeServerResponses(
+            String pql,
+            Map<ServerInstance, byte[]> responseMap, boolean isOfflineTable,
```

Comment:
isOfflineTable is always true. Also, one param per line, so wrap isOfflineTable in this case